src/trie.lisp

```diff
@@ -125,15 +134,23 @@
 (defmethod mapkeys ((fn function) (trie trie))
   (labels ((recursive-fun (trie prefix)
              (declare (type trie trie)
-                      (type string prefix))
-             (let ((new-prefix (concatenate 'string prefix
-                                            (or (let ((k (key trie)))
-                                                  (when k (string k)))
-                                                ""))))
+                      (type list prefix))
+             (let ((new-prefix
+                     (let ((k (key trie)))
+                       (if k
+                           (cons (string k) prefix)
+                           prefix))))
                (mapc (lambda (x) (recursive-fun x new-prefix)) (children trie))
                (when (activep trie)
-                 (funcall fn new-prefix)))))
-    (recursive-fun trie ""))
+                 (let* ((len (reduce #'+ new-prefix :key #'length))
+                        (string (make-string len)))
+                   (loop for p in new-prefix
```

Comment:
Alternatively we can move this let and loop into a FLET function, e.g. "prefix-to-key" (though I don't really like this name... throwing out the idea though)